resharper/resharper-unity/src/Rider/RiderPackageUpdateAvailabilityChecker.cs

```diff
@@ -0,0 +1,127 @@
+            myBoundSettingsStore = applicationWideContextBoundSettingStore.BoundSettingsStore;
+            unitySolutionTracker.IsUnityProjectFolder.WhenTrue(lifetime, lt =>
+            {
+                ShowNotificationIfNeeded(lt);
+                BindToInstallationSettingChange(lt);
+            });
+        }
+        
+        private void BindToInstallationSettingChange(Lifetime lifetime)
+        {
+            var entry = myBoundSettingsStore.Schema.GetScalarEntry((UnitySettings s) => s.AllowRiderUpdateNotifications);
+            myBoundSettingsStore.GetValueProperty<bool>(lifetime, entry, null).Change.Advise_NoAcknowledgement(lifetime, args =>
+            {
+                if (!args.GetNewOrNull()) return;
+                ShowNotificationIfNeeded(lifetime);
+            });
+        }
+
+        private void ShowNotificationIfNeeded(Lifetime lifetime)
+        {
+            if (!myBoundSettingsStore.GetValue((UnitySettings s) => s.AllowRiderUpdateNotifications))
+                return;
+            
+            myPackageManager.IsInitialUpdateFinished.WhenTrue(lifetime, lt =>
+            {
+                var version = myUnityVersion.ActualVersionForSolution.Value;
+                if (version == null)
+                    return;
+            
+                if (myNotificationShown.Contains(mySolution.SolutionFilePath)) return;
+
+                // Version before 2019.2 doesn't have Rider package
+                // 2019.2.0 - 2019.2.5 : version 1.2.1 is the last one
+                // 2019.2.6 - present : 3.0.7+
+                if (version < new Version(2019, 2, 6)) return;
+
+                var package = myPackageManager.GetPackageById(packageId);
+
+                if (package == null)
+                {
+                    myNotificationShown.Add(mySolution.SolutionFilePath);
+                    myLogger.Info($"{packageId} is missing.");
+                    var notification = new NotificationModel($"JetBrains Rider package in Unity is missing.", "Make sure JetBrains Rider package is installed in Unity Package Manager.", true, RdNotificationEntryType.WARN, new List<NotificationHyperlink>());
+                    myShellLocks.ExecuteOrQueueEx(lt, "RiderPackageUpdateAvailabilityChecker.ShowNotificationIfNeeded", () => myNotifications.Notification(notification));
+                }
+                // todo: read available newest compatible Rider package version from somewhere, like protocol or json file
+                else if (package.Source == PackageSource.Registry && new Version(package.PackageDetails.Version) < new Version(3,0,7))
+                {
+                    myNotificationShown.Add(mySolution.SolutionFilePath);
+                    myLogger.Info($"{packageId} {package.PackageDetails.Version} is older then expected.");
+                    var neverShowLink = new NotificationHyperlink("Never show for this solution", myHyperlinkId);
+                    var notificationId = 123409;
```

Comment:
Where does this magic number come from?

Reply:
It is a random number. Any number except conflicting with other notifications would work.
I asked in https://jetbrains.slack.com/archives/CH9S2GGRG/p1637923373479800, it seems it is good enough. But I will evaluate other suggestions.